### PR TITLE
Flax Masked Language Modeling training example

### PR DESCRIPTION
**Pull request by mfuntowicz**
Include a training example running with Flax/JAX framework. (cc @avital @marcvanzee)

TODOs:

- [x] Make the collator working with Numpy/JAX array 
- [x] Make sure the training actually works on larger scale
- [x] Make it possible to train from scratch
- [x] Support TPU (`bfloat16`)
- [ ] Support GPU amp (`float16`)
- [x] Improve overall UX

### CODE REVIEW

File: src/transformers/modeling_flax_bert.py [hunk 1]
```diff
@@ -417,3 +403,104 @@ def __call__(self, input_ids, attention_mask=None, token_type_ids=None, position
...
+        pooled = FlaxBertPooler(name="pooler")(encoder)
+        logits = FlaxBertLMPredictionHead(vocab_size=self.vocab_size, name="predictions")(encoder)
+
+        return pooled, logits
+
+
+class FlaxBertForMaskedLM(FlaxBertModel):
+    def __init__(self, config: BertConfig, state: dict, seed: int = 0, **kwargs):
+        super().__init__(config, state, seed, **kwargs)
+        self.module = partial(
```
Comment (avital) [inline src/transformers/modeling_flax_bert.py:464]: I know I'm late to the party but can you also just return a module instance rather than a partial, e.g. `return FlaxBertOnlyMLMHead(vocab_size=...)`?

Comment (avital) [reply src/transformers/modeling_flax_bert.py:464]: er `self.module = FlaxBertOnlyMLMHead(vocab_size=..., ...)`

File: examples/language-modeling/run_mlm_wwm_flax.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+    for epoch in track(range(int(training_args.num_train_epochs)), description="Training..."):
+        samples_idx = np.random.choice(len(tokenized_datasets["train"]), (training_args.train_batch_size, ))
+        samples = [tokenized_datasets["train"][idx.item()] for idx in samples_idx]
+        model_inputs = data_collator(samples)
+
+        # Model forward
+        # TODO: Remove this conversion by replacing the collator
+        model_inputs = {var_name: tensor.numpy() for var_name, tensor in model_inputs.items()}
+        loss, optimizer = training_step(optimizer, model_inputs)
```
Comment (thomwolf) [inline examples/language-modeling/run_mlm_wwm_flax.py:315]: This is really cool!

Can you also easily have a learning rate schedule?

File: examples/language-modeling/run_mlm_wwm_flax.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    # Setup optimizer
+    optimizer = Adam(
+        learning_rate=training_args.learning_rate,
+        weight_decay=training_args.weight_decay
+    ).create(model.params)
+
+    for epoch in track(range(int(training_args.num_train_epochs)), description="Training..."):
+        samples_idx = np.random.choice(len(tokenized_datasets["train"]), (training_args.train_batch_size, ))
+        samples = [tokenized_datasets["train"][idx.item()] for idx in samples_idx]
+        model_inputs = data_collator(samples)
```
Comment (thomwolf) [inline examples/language-modeling/run_mlm_wwm_flax.py:310]: What's your first impression feeling on having a FLAX Trainer with a similar API to the PT Trainer at some point @sgugger?

Comment (sgugger) [reply examples/language-modeling/run_mlm_wwm_flax.py:310]: Doesn't look like it's going to be too hard to build.

**Review by sgugger — COMMENTED** [review]
Argh, you actually took the one example that is a bit flaky (we merged `DataCollatorForWholeWordMasking` a bit too fast and the data preprocessing part of this script needs to be completely rewritten as it works for BERT only for now).
Could you do the same with the `run_mlm` script instead? This one won't change :-)

File: examples/language-modeling/run_mlm_wwm_flax.py (hunk 2 above)
Comment (mfuntowicz) [reply examples/language-modeling/run_mlm_wwm_flax.py:315]: Yap sure, didn't put it in at first, focusing on making things clear and almost a no-brainer 😄. 

Will look at it soon 👍

File: src/transformers/modeling_flax_bert.py (hunk 1 above)
Comment (mfuntowicz) [reply src/transformers/modeling_flax_bert.py]: Done 👍

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
 import flax.linen as nn
 import jax
 import jax.numpy as jnp
+from functools import partial
```
Comment (marcvanzee) [inline src/transformers/models/bert/modeling_flax_bert.py:23]: This seems to be unused.

File: examples/language-modeling/run_mlm_flax.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+def training_step(optimizer, batch):
+    # def one_hot(labels, num_classes):
+    #     x = labels[..., None] == jnp.arange(num_classes)[None]
+    #     return x.astype(jnp.float32)
+    #
+    # def cross_entropy(logits, labels):
+    #     return -jnp.mean(jnp.sum(one_hot(labels, config.vocab_size) * logits, axis=-1), axis=-1)
+    #
+
+    def cross_entropy(logits, targets, label_smoothing=0.0):
```
Comment (marcvanzee) [inline examples/language-modeling/run_mlm_flax.py:159]: Maybe factor this function out of `training_step` to make it easier to read?

File: examples/language-modeling/run_mlm_flax.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    set_seed(training_args.seed)
+
+    # Get the datasets: you can either provide your own CSV/JSON/TXT training and evaluation files (see below)
+    # or just provide the name of one of the public datasets available on the hub at https://huggingface.co/datasets/
+    # (the dataset will be downloaded automatically from the datasets Hub).
+    #
+    # For CSV/JSON files, this script will use the column called 'text' or the first column if no column called
+    # 'text' is found. You can easily tweak this behavior (see below).
+    #
+    # In distributed training, the load_dataset function guarantee that only one local process can concurrently
```
Comment (marcvanzee) [inline examples/language-modeling/run_mlm_flax.py:247]:
```suggestion
    # In distributed training, the load_dataset function guarantees that only one local process can concurrently
```

File: examples/language-modeling/run_mlm_wwm_flax.py (hunk 2 above)
Comment (mfuntowicz) [reply examples/language-modeling/run_mlm_wwm_flax.py]: Now included 👍
